File: Search/TreeBranchLeafSearch/LeftRightRootHeight/RootHeight.cs
using System;
using System.Collections.Generic;

namespace LeftRightRootHeight
{
    /// <summary>
    /// Find the height of a tree
    /// </summary>
    class RootHeight
    {
        /// <summary>
        /// Main - start point
        /// </summary>
        /// <param name="args"></param>
        static void Main(string[] args)
        {
            // Create inputs
            Node leaf1 = new Node(null, null);
            Node leaf2 = new Node(null, null);
            Node branch = new Node(leaf1, null);
            Node root = new Node(branch, leaf2);

            // Test algorithm
            Console.WriteLine($"Height of a Tree of branch and leaf: {root.Height()}");
            Console.WriteLine($"Height of a branch of 1 leaf: {branch.Height()}");
            Console.WriteLine($"Height of a leaf: {leaf1.Height()}");

            // end
            Console.WriteLine("end!");
            Console.ReadLine();
        }
    }

    /// <summary>
    /// Create a Node object that can only have 1 parent and 2 child: left and right
    /// </summary>
    public class Node
    {
        // Left and right children
        public Node LeftChild { get; private set; }
        public Node RightChild { get; private set; }

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="leftChild"></param>
        /// <param name="rightChild"></param>
        public Node(Node leftChild, Node rightChild)
        {
            this.LeftChild = leftChild;
            this.RightChild = rightChild;
        }

        /// <summary>
        /// Search all paths
        /// </summary>
        /// <param name="pathNodes"></param>
        /// <param name="node"></param>
        /// <param name="pathHeight"></param>
        /// <returns></returns>
        private List<Node> GetPaths(List<Node> pathNodes,
            Node node, ref int pathHeight)
        {
            // if we are done in this path
            if (node == null)
            {
                return pathNodes;
            }

            // search this node
            pathHeight++;
            pathNodes.Add(node);

            // if this is a leaf, no further paths
            if (node.LeftChild == null &&
                node.RightChild == null)
            {
                pathHeight--;
                pathNodes.Remove(node);
                return pathNodes;
            }

            // search left and right
            pathNodes = GetPaths(pathNodes, node.LeftChild, ref pathHeight);
            pathNodes = GetPaths(pathNodes, node.RightChild, ref pathHeight);

            // we are done searching this particular node
            pathNodes.Remove(node);

            return pathNodes;
        }

        /// <summary>
        /// Find the height of a tree
        /// </summary>
        /// <returns></returns>
        public int Height()
        {
            // Initiate the height
            int height = 0;

            // search all paths
            GetPaths(new List<Node>(), this, ref height);

            return height;
        }
    }
}

File: Search/TreeSearch/Javascript/LeftRightRootHeight/rootheight.js
// Create a Node object that can only have 1 parent and 2 child: 
// left and right
class Node
{
    // Constructor
    // define Left and Right children 
    constructor(leftChild, rightChild)
    {
        this.LeftChild = leftChild;
        this.RightChild = rightChild;
    }

    // search all paths
    getPaths(pathNodes, node, pathHeight)
    {
        // if we are done in this path
        if (node == undefined)
        {
            return pathNodes;
        }

        // search this node
        pathHeight.value++;
        pathNodes.push(node);

        // if this is a leaf, no further paths
        if (node.LeftChild == undefined &&
            node.RightChild == undefined)
            {
                pathHeight.value--;
                pathNodes = pathNodes.filter(n => n != node);
                return pathNodes;
            }

        // search left and right
        pathNodes = this.getPaths(pathNodes,
            node.LeftChild, pathHeight);
        pathNodes = this.getPaths(pathNodes,
            node.RightChild, pathHeight);

        // we are done searching this particular node
        pathNodes = pathNodes.filter(n => n != node);

        return pathNodes;
    }

    // Find the height of a tree
    height()
    {
        // Initiate the height
        let path_Height = { value: 0 };

        let nodes = [];

        // search all paths
        this.getPaths(nodes, this, path_Height);

        return path_Height.value;
    }
}

// Main - start point
function main () {
    // create inputs
    leaf1 = new Node(null, null);
    branch = new Node(leaf1, null);
    leaf2 = new Node(null, null);
    root = new Node(branch, leaf2);

    // test algorithm
    console.log("Height of a leaf : " + 
        leaf1.height());
    console.log("Height of a branch of 1 leaf : " + 
        branch.height());
    console.log("Height of a tree of 1 branch and 1 leaf : " + 
        root.height());
}

main()